### DISCUSSION TITLE
File data is not available in the submit data with a Controller and file input

### DISCUSSION THREAD

**Discussion by felix-hubert-eai** [react-hook-form/react-hook-form#5394] [Q&A] [answered]
**Describe the bug**
When using a Controller component rendering a file input, the file data is not sent through the submit function, only the file name.
 
*On version : 7.6.10*

**To Reproduce**
See the codesandbox. Use both file inputs and add a file, then click on the submit button. Notice that the data that's printed out in the console is `{file1: FileList, file2: "C:\fakepath\Screen Shot 2021-05-21 at 9.19.03 AM.png"}`, working well for the first input but only keeping the filename for the Controller input with a custom rendering.

**Codesandbox link (Required)**
https://codesandbox.io/s/dazzling-chatterjee-2gvc6?file=/src/App.js

**Expected behavior**
The file data is sent properly

**Screenshots**
<img width="725" alt="Screen Shot 2021-05-28 at 4 44 37 PM" src="https://user-images.githubusercontent.com/58744586/120040113-02731300-bfd4-11eb-9a61-91c09536c109.png">

**Desktop (please complete the following information):**
- OS: mac
- Browser : chrome

**Additional context**

**Comment by bluebill1049** [comment]
This is due to what you have supplied to `Controller` during onChange. 

```tsx
render={({ filed }) => {
  <Input onChange={(e) => { filed.onChange(e.target.files) }} /> // instead e.target.value
}}
```

**Reply by felix-hubert-eai** [reply]
Hi, thanks for the answer. 
I haven't supplied anything to the Controller's onChange; unless I misunderstood and i need to change it to the code you suggested, in that case I tried it and this will trigger a crash now because the value of the input is not a string anymore: 

`Failed to set the 'value' property on 'HTMLInputElement': This input element accepts a filename, which may only be programmatically set to the empty string.`

**Reply by 4ortytwo** [reply]
@felix-hubert-eai , how did you go about it?

**Reply by felix-hubert-eai** [reply]
I'm still having the issue, couldn't find anything that worked

**Reply by bluebill1049** [reply]
https://codesandbox.io/s/long-sun-nsfbk?file=/src/App.js what about this?

**Reply by bluebill1049** [reply]
Controller is only meant to capture value and validate it.

**Reply by felix-hubert-eai** [reply]
That seems to work! Thank you so much

**Reply by obeka** [reply]
What I understand is that it is not doable with `Controller` component then, right?

**Reply by rppala3** [reply]
Yes it is doable! I solved writing like this
````
<Input
  // {...field} // <- Remove this line
  type="file"
  onChange={(e) => field.onChange(e.target.files?.[0])}
/>
````
This line `{...field}` caused days of pain...

**Reply by harmochiky** [reply]
> Yes it is doable! I solved writing like this
> 
> ```
> <Input
>   // {...field} // <- Remove this line
>   type="file"
>   onChange={(e) => field.onChange(e.target.files?.[0])}
> />
> ```
> 
> This line `{...field}` caused days of pain...

This did it for me

**Comment by bluebill1049** [comment, marked as answer]
https://codesandbox.io/s/long-sun-nsfbk?file=/src/App.js what about this?

**Reply by 4ortytwo** [reply]
Thank you, @bluebill1049!

**Reply by anselm94** [reply]
Thank you!

**Reply by misterflop** [reply]
Thank you, it's really useful! This example would be useful in the documentation!

**Reply by meotimdihia** [reply]
file upload should have an example. I have searched like 10+ topics in Github issues, still have problems:
- I can't use it with  `useFieldArray`.
- How multi-upload works
- How to submit the data.
- How to set default values with Filelist.  
- Why is it behavior different with `<Controller>` in `react-form-hook` and `useController`. This problem is super confusing.
- ....  

I will solve it finally but it'll take days to learn how to do with `react-form-hook`

**Reply by mustakimkr** [reply]
Thank you so much!

**Reply by alizaeda** [reply]
Not working for me

**Reply by asya-aisa** [reply]
Thank you!

**Reply by arimendelow** [reply]
Note that if you try to set value on the input like that in TypeScript, you'll get this:

<img width="924" alt="Screenshot 2023-11-26 at 17 40 33" src="https://github.com/react-hook-form/react-hook-form/assets/16390116/c1f58dc1-2997-4202-b316-aadda3930303">
